content/reference/routing.md

```diff
@@ -0,0 +1,316 @@
+as: `/:user/:repo/blob/:branch/*`. The value of `state.params.wildcard` would
+then be: `'/html/raw.js'`.
+
+Try and use wildcards sparingly. They're the most powerful tool in the routing
+toolbox, which means that if you're not careful you might end up reimplementing
+a router on top of it yourself.
+
+### 404s
+There is one last thing we should touch on with dynamic routing: what to do
+when a route is not found.
+
+If you're using params or wildcards, there can always be a case where a route
+isn't found. For example if we have the route `/:users`, if a particular user
+does not exist, we might want to show a fallback route instead.
+
+There are generally two approaches to this: hard redirects, and soft redirects.
+- A hard redirect is when we redirect to a new route. For example if `/foobar`
+  doesn't match a known user, we'll navigate to `/404` instead.
+- A soft redirect is when the URL is kept the same, but different content is
+  shown. For example if `/foobar` doesn't match, we'll require and render the
+  content of the `/404` view instead.
+
+It's generally recommended to use soft redirects, as they interfere the least
+with the browser, and allow users to recover from an error (e.g. fix typos in a
+url).
+
+### Navigating to External Links
+Some links in your app will point to other pages. In order to do this safely,
+we must add some attributes to our link tags. This is needed, so the pages we
+link to can't hijack our page.
+
+In order to link to an external link, we must do one of the following:
+- the click event had `.preventDefault()` called on it
+- the link has a `target="_blank"` attribute with `rel="noopener noreferrer"`
+- a modifier key is enabled (e.g. `ctrl`, `alt`, `shift` or `meta`)
+- the link's href starts with protocol handler such as `mailto:` or `dat:`
+- the link points to a different host
+- the link has a `download` attribute
+
+```js
+var html = require('choo/html')
+html`
+  <a href="other.site.com/something" target=_blank rel="noopener noreferrer">
+    Click to open in another page
+  </a>
+`
+```
+
+### Programmatic Navigation
+Often it's needed to change routes after some event happens. For example,
+someone logs in, and we need to redirect them to the logged in page. We need
+programmatic navigation.
+
+Choo comes with several events built-in to allow you to navigate using the
+Browser's History API. The History API is a thin wrapper around a
+[Stack](https://en.wikipedia.org/wiki/Stack_(abstract_data_type)). As you
+navigate, history entries are added to the stack. And when you hit the back
+button, it pops entries off the stack.
+
+- `'pushState'` - navigate to a new route, and push an entry to the history
+  stack.
+- `'replaceState'` - navigate to a new route, and replace the current entry
+  on the history stack with a new one.
+- `'popState'` - pop an entry off the history stack, and navigate to a
+  previous route.
+- `'navigate'` - emitted when any of the above events is emitted.
+
```

Comment:
this list is good, but it'd be better if it recommended usage cases, such as the fact that you should commonly use `pushState` instead of `replaceState`